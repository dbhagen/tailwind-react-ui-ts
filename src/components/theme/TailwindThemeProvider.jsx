import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
