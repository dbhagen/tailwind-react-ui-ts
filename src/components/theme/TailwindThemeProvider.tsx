import React from 'react'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

export type TailwindThemeProviderProps = {
  theme: {}
  children: React.ReactNode
}

const TailwindThemeProvider: React.FunctionComponent<TailwindThemeProviderProps> = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

export default TailwindThemeProvider
