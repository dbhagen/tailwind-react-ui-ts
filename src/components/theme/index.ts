export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
