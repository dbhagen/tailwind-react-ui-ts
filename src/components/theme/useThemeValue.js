import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
