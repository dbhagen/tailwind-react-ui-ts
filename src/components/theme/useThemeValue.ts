import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, values: Object, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(undefined, values, prefix)
}
