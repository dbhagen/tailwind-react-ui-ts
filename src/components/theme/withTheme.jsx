import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
