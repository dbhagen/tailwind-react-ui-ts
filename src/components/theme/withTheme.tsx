import React from 'react'

import TailwindTheme from './TailwindTheme'

// TODO: Remove Any Types
const withTheme = (Component: any) => {
  const WithTheme = (props: any) => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
