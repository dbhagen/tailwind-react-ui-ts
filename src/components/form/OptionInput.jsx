import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { Touchable, Flex, Box } from '../primitives'
import { withTheme } from '../theme'

import Label from './Label'

class OptionInput extends PureComponent {
  constructor(props) {
    super(props)

    const { defaultChecked, checked } = props

    this.state = {
      checked: defaultChecked || !!checked,
    }

    this.inputRef = null

    this.handleChange = this.handleChange.bind(this)
    this.handleUpdate = this.handleUpdate.bind(this)
  }

  componentDidUpdate(prevProps) {
    this.handleUpdate(prevProps)
  }

  handleChange(e) {
    const { checkbox, onChange, checked } = this.props

    if (checked === undefined) {
      this.setState(({ checked: checkedState }) => ({
        checked: checkbox ? !checkedState : true,
      }))
    }

    if (onChange) onChange({ ...e, target: this.inputRef.current })

    e.preventDefault()
  }

  handleUpdate() {
    const { checked } = this.props
    const { checked: checkedState } = this.state

    if (checked !== undefined && checked !== checkedState) {
      this.setState({ checked })
    }
  }

  render() {
    const {
      theme,
      name,
      value,
      label,
      hideLabel,
      checkbox,
      id,
      ...rest
    } = this.props
    const { checked } = this.state

    this.inputRef = React.createRef()

    return (
      <Touchable
        is={Label}
        flex
        items="center"
        onTouch={this.handleChange}
        {...rest}
      >
        <Box
          is="input"
          id={id}
          visuallyHidden
          name={name}
          type={checkbox ? 'checkbox' : 'radio'}
          value={value}
          checked={checked}
          tabIndex="-1"
          innerRef={this.inputRef}
          onChange={() => {}}
        />
        <Flex
          items="center"
          justify="center"
          inlineBlock
          rounded={checkbox ? theme.radius : 'full'}
          h={4}
          w={4}
          border={[true, checked ? theme.brandColors.primary : false].filter(
            prop => !!prop,
          )}
          bg={checkbox && checked ? theme.brandColors.primary : undefined}
          m={{ r: theme.spacing.sm }}
        >
          {checked &&
            (checkbox ? (
              <Box
                is="svg"
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 24 24"
                h={3}
                w={3}
                text="white"
                fill="current"
              >
                <path d="M0 0h24v24H0z" fill="none" />
                <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
              </Box>
            ) : (
              <Box
                inlineBlock
                rounded="full"
                h={2}
                w={2}
                bg={theme.brandColors.primary}
              />
            ))}
        </Flex>
        <Box
          inlineBlock
          visuallyHidden={hideLabel}
          leading="tight"
          font="normal"
        >
          {label}
        </Box>
      </Touchable>
    )
  }
}

OptionInput.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  hideLabel: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  checked: PropTypes.bool,
  onChange: PropTypes.func,
  checkbox: PropTypes.bool,
  id: PropTypes.string,
}

OptionInput.defaultProps = {
  hideLabel: false,
  defaultChecked: undefined,
  checked: undefined,
  onChange: undefined,
  checkbox: false,
  id: undefined,
}

export { OptionInput as component }
export default withTheme(OptionInput)
