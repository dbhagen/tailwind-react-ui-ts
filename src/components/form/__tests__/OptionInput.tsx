import React from 'react'
import { shallow } from 'enzyme'

import { component as OptionInput } from '../OptionInput'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign(
    { theme: defaultTheme, name: 'test', value: 'test-1', label: 'Test?' },
    testProps,
  )

  const wrapper = shallow(<OptionInput {...props} />)

  return {
    props,
    wrapper,
  }
}

describe('OptionInput', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup({ defaultChecked: true })

    expect(wrapper).toMatchSnapshot()
  })
})
