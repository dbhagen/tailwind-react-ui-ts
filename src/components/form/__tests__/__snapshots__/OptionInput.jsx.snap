// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionInput renders matching snapshot 1`] = `
<Touchable
  defaultChecked={true}
  disabled={false}
  flex={true}
  is={[Function]}
  items="center"
  onTouch={[Function]}
>
  <Box
    checked={true}
    inline={false}
    inlineBlock={false}
    innerRef={
      Object {
        "current": null,
      }
    }
    is="input"
    name="test"
    onChange={[Function]}
    tabIndex="-1"
    type="radio"
    value="test-1"
    visuallyHidden={true}
  />
  <Flex
    border={
      Array [
        true,
        "blue-400",
      ]
    }
    col={false}
    h={4}
    inline={false}
    inlineBlock={true}
    inlineFlex={false}
    is="div"
    items="center"
    justify="center"
    m={
      Object {
        "r": 2,
      }
    }
    reverse={false}
    rounded="full"
    w={4}
    wrap={false}
    wrapReverse={false}
  >
    <Box
      bg="blue-400"
      h={2}
      inline={false}
      inlineBlock={true}
      is="div"
      rounded="full"
      w={2}
    />
  </Flex>
  <Box
    font="normal"
    inline={false}
    inlineBlock={true}
    is="div"
    leading="tight"
    visuallyHidden={false}
  >
    Test?
  </Box>
</Touchable>
`;
